hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -17,210 +17,212 @@
         assert not self.is_exhausted
-        prefix = bytearray()
-        node = 0
+
+        result = bytearray()
+
+        def append_int(n_bits, value):
+            result.extend(int_to_bytes(value, bits_to_bytes(n_bits)))
+
+        current_node = self.root
+
         while True:
-            assert len(prefix) < self.cap
-            assert node not in self.dead
-
-            # Figure out the range of byte values we should be trying.
-            # Normally this will be 0-255, unless the current position has a
-            # mask.
-            mask = self.masks.get(node, 0xFF)
-            assert _is_simple_mask(mask)
-            upper_bound = mask + 1
-
-            try:
-                c = self.forced[node]
-                # This position has a forced byte value, so trying a different
-                # value wouldn't be helpful. Just add the forced byte, and
-                # move on to the next position.
-                prefix.append(c)
-                node = self.nodes[node][c]
-                continue
-            except KeyError:
-                pass
-
-            # Provisionally choose the next byte value.
-            # This will change later if we find that it was a bad choice.
-            c = random.randrange(0, upper_bound)
-
-            try:
-                next_node = self.nodes[node][c]
-                if next_node in self.dead:
-                    # Whoops, the byte value we chose for this position has
-                    # already been fully explored. Let's pick a new value, and
-                    # this time choose a value that's definitely still alive.
-                    choices = [
-                        b
-                        for b in hrange(upper_bound)
-                        if self.nodes[node].get(b) not in self.dead
-                    ]
-                    assert choices
-                    c = random.choice(choices)
-                    node = self.nodes[node][c]
-                else:
-                    # The byte value we chose is in the tree, but it still has
-                    # some unexplored descendants, so it's a valid choice.
-                    node = next_node
-                prefix.append(c)
-            except KeyError:
-                # The byte value we chose isn't in the tree at this position,
-                # which means we've successfully found a novel prefix.
-                prefix.append(c)
+            assert not current_node.exhausted
+            for i, (n_bits, value) in enumerate(
+                zip(current_node.bit_lengths, current_node.values)
+            ):
+                if i in current_node.forced:
+                    append_int(n_bits, value)
+                    continue
+                while True:
+                    v = random.getrandbits(n_bits)
+                    if v != value:
+                        append_int(n_bits, v)
+                        return hbytes(result)
```

Comment:
I notice that when this loop accidentally generates the current value, it will keep spinning until it manages to generate a different one.

That will cause it to aggressively split at the first non-forced position in a node, with no possibility of advancing to the next position. So the surrounding for-loop only serves to advance past any leading forced positions.

Is this unintentional? It's inconsistent with how branches are handled.

Reply:
Hmm. It was intentional but now that you point it out, I agree it's not ideal. I'll write a different implementation of this, thanks.

Reply:
OK, I've replaced it with an implementation that generates the prefix purely uniformly at random (with an optimisation that is required to make that efficient)